### PR TITLE
CRv2: Update attribute processing methods

### PR DESCRIPTION
**Pull request by hacodeorg**
[PLC-918](https://codedotorg.atlassian.net/browse/PLC-918): Updating methods to process contact `opt_in` and `updated_at` attributes after we removed the unique email-sources constraint from ContactRollupsRaw in https://github.com/code-dot-org/code-dot-org/pull/35285 and added 12 more raw data extraction methods in https://github.com/code-dot-org/code-dot-org/pull/35347.

## Testing story
- Ran local unit test
- Running full pipeline on an adhoc machine with production-cloned db (_to be updated_)

### CODE REVIEW

File: dashboard/app/models/contact_rollups_processed.rb [hunk 1]
```diff
@@ -92,11 +92,28 @@ def self.get_data_aggregation_query
...
-  # Parses a JSON string contains all data and metadata of a contact.
-  # It will throw exception if cannot parse the entire input string.
+  # Parses a JSON string containing contact data and metadata.
+  # Throw exception if cannot parse the entire input string.
   #
-  # @param [String] str represents a JSON array. Each array item is a hash {sources:String, data:Hash, data_updated_at:DateTime}.
-  # @return [Hash] a hash with string keys {table_name => {field_name => value}}
+  # @example
+  #   Input string:
+  #     '[{"sources":"table1", "data":{"opt_in":1}, "data_updated_at":"2020-06-16"}]'
```
Comment (bencodeorg) [inline dashboard/app/models/contact_rollups_processed.rb]: Is the input string actually:

`[{"s":"table1", "d":{"opt_in":1}, "u":"2020-06-16"}]`

File: dashboard/app/models/contact_rollups_processed.rb [hunk 2]
```diff
@@ -92,11 +92,28 @@ def self.get_data_aggregation_query
...
+  #   Output hash:
+  #     {'table1'=>{
+  #       'opt_in'=>[{'value'=>1, 'data_updated_at'=>2020-06-16}],
+  #       'last_data_updated_at'=>2020-06-6
```
Comment (bencodeorg) [inline dashboard/app/models/contact_rollups_processed.rb]: Typo I think

File: dashboard/test/models/contact_rollups_processed_test.rb [hunk 3]
```diff
@@ -123,35 +146,82 @@ class ContactRollupsProcessedTest < ActiveSupport::TestCase
...
+            'last_data_updated_at' => time_parsed
+          }
+        }
+      },
+      # input data has more than one value for a key
+      {
+        input: format(
+          '['\
+          '{"%{sources_key}": "table2", "%{data_key}": {"state": "WA"}, "%{data_updated_at_key}": "%{time_str}"},'\
+          '{"%{sources_key}": "table2", "%{data_key}": {"state": "OR"}, "%{data_updated_at_key}": "%{time_str}"}'\
```
Comment (bencodeorg) [inline dashboard/test/models/contact_rollups_processed_test.rb]: Super impressive testing in general, but curious if there's anything that checks that we produce the correct `last_data_updated_at` when input data has  more than one value for a key?

**Review by bencodeorg — APPROVED** [review]
👏 wow this is impressive, I think it looks good! I know owe you an explanation of what I find confusing about the processing step, but I guess my general feeling is any time I review process step code my brain feels like it's in overdrive. That may be unavoidable (it's just a complicated step), and I think is probably exacerbated by removing the unique constraint on email-source (which I think you knew would happen :).

File: dashboard/app/models/contact_rollups_processed.rb (hunk 1 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_processed.rb]: That's right. I'll explain that in the comment

File: dashboard/app/models/contact_rollups_processed.rb (hunk 2 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_processed.rb]: Fixed.

File: dashboard/test/models/contact_rollups_processed_test.rb (hunk 3 above)
Comment (hacodeorg) [reply dashboard/test/models/contact_rollups_processed_test.rb]: Yes good catch, I will add that. I'm also thinking about breaking this long test into smaller ones to make it a bit clearer.

**Pull request merged**